File: Offer/2254/main.py
from typing import Tuple


class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if root == None:
            return None

        temp1, temp2 = None, None

        def dfs(node) -> Tuple:
            if node.left == None and node.right == None:
                return node, node

            nonlocal temp1, temp2

            if node.right == None:
                temp1 = dfs(node.left)
                node.left, temp1[1].right = temp1[1], node
                return temp1[0], node

            if node.left == None:
                temp2 = dfs(node.right)
                node.right, temp2[0].left = temp2[0], node
                return node, temp2[1]

            temp1, temp2 = dfs(node.left), dfs(node.right)
            node.left, temp1[1].right = temp1[1], node
            node.right, temp2[0].left = temp2[0], node
            return temp1[0], temp2[1]

        temp1 = dfs(root)
        temp1[0].left, temp1[1].right = temp1[1], temp1[0]
        return temp1[0]

File: Offer/2254/main.go
package main

type Node struct {
	Val   int
	Left  *Node
	Right *Node
}

func treeToDoublyList(root *Node) *Node {
	if root == nil {
		return nil
	}

	var l1, l2, r1, r2 *Node

	var dfs func(*Node) (*Node, *Node)
	dfs = func(node *Node) (*Node, *Node) {
		if node == nil {
			return nil, nil
		}
		if node.Left == nil && node.Right == nil {
			return node, node
		}
		if node.Right == nil {
			l1, l2 = dfs(node.Left)
			node.Left, l2.Right = l2, node
			return l1, node
		}
		if node.Left == nil {
			r1, r2 = dfs(node.Right)
			node.Right, r1.Left = r1, node
			return node, r2
		}
		l1, l2 = dfs(node.Left)
		node.Left, l2.Right = l2, node
		r1, r2 = dfs(node.Right)
		node.Right, r1.Left = r1, node
		return l1, r2
	}
	l1, r2 = dfs(root)
	l1.Left, r2.Right = r2, l1
	return l1
}
